clang/lib/Sema/SemaDecl.cpp

```diff
@@ -15845,8 +15845,10 @@ void Sema::CheckCoroutineWrapper(FunctionDecl *FD) {
   RecordDecl *RD = FD->getReturnType()->getAsRecordDecl();
   if (!RD || !RD->getUnderlyingDecl()->hasAttr<CoroReturnTypeAttr>())
     return;
-  // Allow `get_return_object()`.
-  if (FD->getDeclName().isIdentifier() &&
+  // Allow some_promise_type::get_return_object().
+  // Since we are still in the promise definition, we can only do this
+  // heuristically as the promise may not be yet associated to a coroutine.
+  if (isa<CXXMethodDecl>(FD) && FD->getDeclName().isIdentifier() &&
       FD->getName().equals("get_return_object") && FD->param_empty())
     return;
```

Comment:
Can we return directly if it has `CoroDisableLifetimeBoundAttr`? (Maybe we need to edit the semantics of `CoroDisableLifetimeBoundAttr` a little bit)

Reply:
I don't understand what purpose it would serve. Do you want `[[clang::coro_disable_lifetimebound]]` to imply `[[clang::coro_wrapper]]` for brevity?

Reply:
My intention here is to remove the special handling to 'get_return_object'. The 'coro_wrapper'  attribute should be applied here IIUC.

Reply:
At this point (in `ActOnFinishFunctionBody`), neither `CoroDisableLifetimeBoundAttr` nor `CoroWrapperAttr` would be available since these are only built once we encounter a coroutine body.

Reply:
Got it. In this case, we can make it into a PendingXXX struct in Sema and we can check it in the end of the TU. But this is not required in this patch. If you don't want to address that in the current patch, let's leave a FIXME or a TODO here.

Reply:
I do not think that would work either. For example, consider TUs with only promise definitions and no coroutine definitions. These attr would not be available even in the end of TU.

Reply:
> I do not think that would work either. For example, consider TUs with only promise definitions and no coroutine definitions. These attr would not be available even in the end of TU.

But in that case we shouldn't see any `coro_return_type` annotations either and there will be no pending actions? I don't see why this approach won't work.

I am worried about breaking the ODR rules, though. @ChuanqiXu9 is it fine to add attributes to existing functions conditioned on how they are used rather than how they are defined?

Imagine two modules:
- one only defines the promise type, but never actually uses it as a coroutine, so we don't happen to get those attributes.
- another one imports the module from the previous step and attributes get added.

We can now easily get object files that have both definitions of those functions with and without the attributes (by importing either one of the other). I am not sure if there are some potential failures lurking in if we do that, e.g. could ODR hash checks for importing both as submodules will fail?

Reply:
> But in that case we shouldn't see any coro_return_type annotations either and there will be no pending actions? I don't see why this approach won't work.

Why not ? We would always see the `coro_return_type` annotation for the return type. The declaration (and therefore the annotations) of `CoroReturnType` would be visible to `CoroReturnType get_return_object() { return {}; }`. This TU does not have to have a coroutine definition.

Reply:
Ah, right, sorry. I somehow thought we only warn when there are real coroutine somewhere, but the whole point of this check is to flag non-coroutine functions that are not flagged as wrappers.

Reply:
> I do not think that would work either. For example, consider TUs with only promise definitions and no coroutine definitions. These attr would not be available even in the end of TU.

Oh, yeah. I didn't think about modules : ).

>  @ChuanqiXu9 is it fine to add attributes to existing functions conditioned on how they are used rather than how they are defined?

Then it is troublesome to use this attribute, which violates our intention.

Reply:
Thanks @ilya-biryukov for flagging the issue with Modules. Then I don't think it is wise to add these attributes. I will revert to the not-so-great heuristic of matching function names.